revise RFD 153 for scoped resources (#68436)
gravitational/teleport: rfd/0153-resource-guidelines.md

- [ ] Create proto spec for resource and RPC service
- [ ] Create backend service
- [ ] Implement gRPC service with authz
- [ ] Add resource client to `api` client
- [ ] Add resource support to tctl (get, create, edit)
- [ ] Add resource parser to event stream mechanism
- [ ] Optional: Add resource to cache
- [ ] Add support for bootstrapping the resource
- [ ] Add support for resource to Teleport Operator
- [ ] Add support for resource to Teleport Terraform Provider

### Defining a resource

At the top level, all new resources MUST include:

- `kind`: a string indicating the kind of resource

> - `sub_kind`: a string indicating the resource sub_kind, the field must be defined and validated even if it is unused.

Reviewer: If it's unused, what do we set it to? Just validate an empty string all the time?

Reviewer 2: ```suggestion
- `sub_kind`: a string indicating the resource sub_kind, the field must be validated to be a known value. If no subkinds are planned to be used at the time of the definition of the resource, a blank string should be used and the validation should check that the sub_kind field is an empty string, or it would be impossible to ever use the field in the future without older binaries potentially misinterpreting resources.
```

- `version`: a string indicating the resource version
- `metadata`: a `teleport.header.v1.Metadata` including the resource name, labels, description, expiration, and revision.
- `spec`: a message including all other user-specified resource fields.

Most new resources SHOULD include:

> - `scope`: a string indicating the administrative scope of the resource, see RFD 229.

Reviewer 3: Imo - we should make life easier for the reader here, and provide a little more context rather than just directing them to 229. I think "most" (unqualified by a condition) is a little strong when there's a lot of ongoing work outside of scopes at the moment.

Reviewer 4: +1.

How do I choose whether to add a scope or not? Should I default to having a scope?

Author: Ah but to answer these questions for the document I will have to answer them for myself. I can agree that an unqualified "most" is probably too strong. What I really want to say is "resources should include a scope unless it makes no sense for them to include a scope, if you don't know what that means ask someone who does"

Reviewer 4: I still think some context or guideline could be very helpful.

Let's try the opposite, when do I not need a scope? If my resource is global I assume not. What other examples?

New resources MAY include:

> - `status`: a message including all resource state that is computed server-side and not user-provided.

Reviewer 5: we should enforce that the status field must be stored under a different key. With the issue with bound key pair token, we must enforce this for every single resource and build a way of easily storing status under <key>/status to prevent users from replacing and overriding teleport managed status fields.

Since we ar enforcing new rules, lets just make them first class citizens

Reviewer 4: > prevent users from replacing and overriding teleport managed status fields

I think a guideline saying we should (must?) validate the status and avoid unintended user-made changes could be enough.

Deviating from Teleport storage expectations, like unary resource names or a single storage entry per resource, usually causes a lot of headaches. For example, if one chooses to save status to a separate key then using generic.Service automatically becomes a poor choice, as it won't do atomic spec+status writes.

Reviewer 5: that’s my goal here: to extend the behavior. Objects created with `tctl create -f`, or applied through IaC tools such as the Operator or Terraform, will not include the status field, so they should be accepted and processed normally

otherwise, we would need to read the existing object and manually preserve or restore its status before every update. that would be error-prone and would depend on each person writing or reviewing the code remembering to do it correctly

we should make this work by design. From my perspective, the status should be stored under a separate key and updated through dedicated calls, similar to how Kubernetes handles resource status

Reviewer 4: I don't disagree, but I would like to see our examples and internal libraries changed to support that as well.

While the kind and version may seem like they would be easy to derive from the message definition itself, they need to be defined so that anything processing a generic resource can identify which resource is being processed. For example, `tctl` interacts with resources in their in raw yaml text form and leverages `services.UnknownResource` to identify the resource and act appropriately.

All properties defined outside the `status` of a resource MUST only be modified by the
owner/creator of the resource. For example, if a resource is created via
`tctl create`, then any fields within the `spec` MUST not be altered dynamically
by the Teleport process. When Teleport automatically modifies the `spec` during
runtime it causes drift between what is in the Teleport backend and the state stored by external IaC tools.
If a resource has properties that must be modified by Teleport, a separate `status` field should
be added to the resource to contain them. These fields will be ignored by IaC tools during their reconciliation.

Do NOT define a `CheckAndSetDefaults()` method for new resource kinds, default
values outside of the resource `status` must not be set.

> Resources MUST have a `Validate` or `StrongValidate` function that applies

Reviewer 4: I don't think a public function should be a requirement, in fact I think it should be discouraged. The storage implementation should validate before persisting, but our generic.Service already has a hook for that.

Author: i can make it unexported, makes no difference to me

Reviewer 4: I'll take the unexported func, then. :D

Reviewer 2: What's the advantage of discouraging exported functions for validation? I could see them being useful for dry runs, client-side "live" validation in editors, and the likes.

Reviewer 4: We have a poor history of CheckAndSetDefaults and over-validation. There's also the possibility of drift between client and server validation logic, which can be harmful.

I'm not saying never export your validation functions, if you have strong reason then by all means do it. I'm saying let's not make that the default.

strict validation to a resource whenever it is accepted from an untrusted
source and before it is persisted to storage.

> Resources MAY have a `WeakValidate` function that applies a weaker form of

Reviewer 3: Distinction between Strong/Weak may be challenging for those unfamiliar. I'd suggest we provide some kind of example here of the kinds of fields that are validated weakly (and why we weakly validate them).

> Resources MAY have a `WeakValidate` function that applies a weaker form of

Reviewer 4: I'm not following the idea of a WeakValidate.

If the source is trusted, why is validation needed? Conversely, if validation is needed (eg: we are about to persist), then why wouldn't it be "strongly validated"?

Do we even need to make a ruling on this?

Author: it can be useful in case the resource is completely unusable or insecure in some state, we don't know if the backend has been modified or the cluster has been upgraded and downgraded. There are some existing WeakValidate functions you can refer to. But I'll probably just remove the mention from this document

Reviewer 2: WeakValidate should assert the presence and well formedness of the bits of the resource that make up the backend key because if those are missing we _really_ don't understand what is going on anymore, but anything else should probably not cause an unmarshaling failure.

  string version = 3;
  // Common metadata that all resources share.
  teleport.header.v1.Metadata metadata = 4;

>   // The scope of the resource

Reviewer 4: Maybe:

```suggestion
  // The scope of the resource, as per RFD XXXX.
```

>   string scope = 5;

Reviewer 4: Before we plaster this all over, should we consider a `Scope` message instead?

(Is it already too late?)

Author: I don't understand the ask with no justification, it would be a very big change, it's almost certainly too late to be honest

Reviewer 4: > it's almost certainly too late to be honest

Fair enough.

> The `Update` RPC takes a resource to be updated and must also return the updated resource so that any fields that are

Reviewer 3: ```suggestion
The `Update` RPC takes a resource to be updated and MUST also return the updated resource so that any fields that are
```

#### Upsert

> The `Create` and `Update` RPCs should be preferred over `Upsert` for user-driven interactions,
> see [#1326](https://github.com/gravitational/teleport/issues/1326) for more details.

> > The `Upsert` method is necessary for IaC integrations that define the source of truth for the resource.

Reviewer 4: `tctl create -f`, sadly, often needs to be an Upsert too.

Reviewer 6: I would consider `tctl` as a real IaC tool, exactly like Terraform.

The simplest IaC setup is writing a bunch of manifests in a git repo and running in the CI:

```bash
set -euxo pipefail

for file in *; do 
    if [ -f "$file" ]; then 
        tctl create -f "$file"
    fi 
done
```

In this case, the user does want to be authoritative and Upsert is the adequate behaviour. I think `tctl create -f` should always upsert.

However I agree with you about `tctl` missing a proper `tctl update` command. Many resources also do not implement a sane `Update` RPC and make `tctl edit` vulnerable to races.

additional resources, the response MUST also include a token that indicates where the next page of results begins.

`List` RPC can optionally provide a `Filter` message and/or a `SortMode` field where supported.
It MUST accept a scope filter if the resource is scoped.

> Do not provide an unpaginated `GetAllFoos` RPC.

Reviewer 4: 🙏

`lib/services/local/foo.go`. An accompanying interface which mirrors the service is defined in `lib/services/foo.go`.
Continuing on with the example above, the sections below show how the backend service for the `Foo` resource might look like.

For most cases, when adding a new resource, it is preferred to create a service
that wraps the `generic.ServiceWrapper` or `generic.ScopeAwareServiceWrapper`
over implementing everything from scratch.
If custom behavior is required for a subset of backend operations, they may be
implemented directly while all the other operations still make use of the
generic service.

> If the resource has a scope, it should be namespace by scope, meaning that each

Reviewer: ```suggestion
If the resource has a scope, it should be namespaced by scope, meaning that each
```

resource should be uniquely identified by a `(name, scope)` pair with a unique
backend key.
In this case, the backend service should wrap `generic.ScopeAwareServiceWrapper`.

```go
// FooService is a storage service for Foos.
type FooService struct {
	service *generic.ScopeAwareServiceWrapper[*foov1.Foo]
}

// NewFooService returns a new service for Foos.
func NewFooService(bk backend.Backend) (*FooService, error) {
	service, err := generic.NewScopeAwareServiceWrapper(generic.ScopeAwareServiceWrapperConfig[*foov1.Foo]{
		Backend:               bk,
		ResourceKind:          foos.Kind,
		UnscopedBackendPrefix: backend.NewKey("foo"),
		ScopedBackendPrefix:   backend.NewKey("scoped", "foo"),
		MarshalFunc:           services.MarshalProtoResource[*foov1.Foo],
		UnmarshalFunc:         services.UnmarshalProtoResource[*foov1.Foo],

> 		ValidateFunc:          foos.StrongValidate,

Reviewer 4: My 2c is that this should be a private function to discourage over-validation of resources. Ie, having `tctl` call StrongValidate directly, or having it called in unexpected places in the codebase.

Reviewer 4: I'll resolve in favor of https://github.com/gravitational/teleport/pull/68436#discussion_r3544116197, which is rather similar. Feel free to reopen if this will help you remember about this line.

	})
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return &FooService{
		service: service,
	}, nil
}
```

#### Resource validation

> The strictest validation of a resource should be performed prior to write operations. Any resource persisted in the

Reviewer 3: Now that the document defines `StrongValidate`/`WeakValidate` pattern above, I would make explicit reference to those names here within Resource Validation.

```go
func (s *FooService) UpsertFoo(ctx context.Context, foo *foov1.Foo) (*foov1.Foo, error) {
	return s.service.UpsertResource(ctx, foo)
}
```

#### Get

The Get method should accept the protobuf `Get<Resource>Request` message as a
parameter, so that options can be added to the request in a backward-compatible
manner.

`trace.NotFound` errors from the backend should be re-wrapped to include the resource name instead of the backend key for friendlier user-facing errors.
`generic.(ScopeAware)ServiceWrapper` handles this.

```go
// GetFoo returns a single Foo matching the request
func (s *FooService) GetFoo(ctx context.Context, req *foov1.GetFooRequest) (*foov1.Foo, error) {
	return s.service.GetResource(ctx, scopes.QualifiedName{
		Scope: req.GetScope(),

> 		Name:  req.GetName(),

Reviewer 4: How does this work with composite names? Say "/cert_authority_override/{cluster}/{ca_type}"?

(If the APIs accepted a backend.Key we wouldn't have limitations.)

Author: I'm not trying to propose a full new API for the generic resource backend here, with the current one you can just do `s.service.WithPrefix(clusterName).GetResource(ctx, caType)` (re-wrap the not found error if you want to show `{ca_type}/{cluster}`)

the ScopeAwareService has similar WithScopePrefix/WithScopedResourcePrefix helpers, and there's always `WithNameKeyFunc` which nobody loves but is sometimes necessary when working with the generic service implementations

Reviewer 4: I do think it would be nice if we could support composite keys as a first-class citizen, instead of forcing impls to parse strings, mess around with WithNameKeyFunc, etc.

I understand if you want to push back as this being outside of scope, although given the scopes.QualifiedName definition it'll always be.

func (s *FooService) DeleteFoo(ctx context.Context, req *foov1.DeleteFooRequest) error {
	return s.service.DeleteResource(ctx, scopes.QualifiedName{
		Scope: req.GetScope(),
		Name:  req.GetName(),
	})
}
```

#### List

Listing can be done via collecting a stream returned by
`backend.Backend.Items`, making use of the functional helpers in the
[stream](https://github.com/gravitational/teleport/blob/a32c69b581ff0ceef55b83a601cfb54c6d52d710/lib/itertools/stream/stream.go) package.

The List method should accept the protobuf `List<Resource>Request` message as a parameter, so that options including filters can be added to the request in a backward-compatible manner.

A listing operation should not abort entirely if a single item cannot be
converted from a `backend.Item`, it should instead be logged, and the rest of
the page should be processed.
Aborting an entire page when a single entry is invalid causes Teleport to be
permanently unhealthy since it is never able to load or cache the affected
resource(s).
`generic.(ScopeAware)ServiceWrapper` handles this.

```go
// ListFoos returns a page of Foos and the token to find the next page of items.

> func (s *FooService) ListFoos(ctx context.Context, req *foov1.ListFoosRequest) ([]*foov1.Foo, string, error) {

Reviewer 2: The change to this signature will require so much boilerplate everywhere that's using a list RPC just to use the helpers (like clientutils.Resources) that we made specifically to reduce boilerplate.

Author: I thought we made `clientutils.Resources` so people would stop rolling their own broken pagination loops. Anyways, if it can't accommodate a filter, has it failed to serve its purpose? Every List API I can find that does accept a filter already accepts the proto request. ListResources, ListUnifiedResources, ListAccessListsV2, ListAccessMonitoringRulesWithFilter, ListNotifications, ListScopedTokens, ListUsers, ListAccessRequests, etc...

This RFD also already does _not_ recommend wrapping the gRPC client, rather it explicitly recommends returning the plain gRPC client https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md#api-client.  Here the List method signature will be `ListFoos(context.Context, *foov1.ListFoosRequest, opts ...grpc.CallOption) (*foov1.ListFoosResponse, error)`, so already how is one meant to use clientutils.Resources without a touch of boilerplate?

Could also add a helper to match this signature without too much difficulty
```
type PageRequest interface {
	proto.Message
	SetPageSize(int32)
	SetPageToken(string)
}

// FilteredResources returns an iterator over all resources from every page produced from the pageFunc.
// The iterator will only produce an error if one is encountered retrieving a page.
func FilteredResources[T any, R PageRequest](
	ctx context.Context,
	pageFunc func(context.Context, R) ([]T, string, error),
	req R,
) iter.Seq2[T, error] {
	req = proto.CloneOf(req)
	return rangeInternal(ctx, rangeParams[T]{
		pageFunc: func(ctx context.Context, pageSize int, pageToken string) ([]T, string, error) {
			req.SetPageSize(int32(pageSize))
			req.SetPageToken(pageToken)
			return pageFunc(ctx, req)
		},
		pageSize: defaults.DefaultChunkSize,
	})
}
```

Reviewer 2: (only messages with hybrid/opaque codegen will have the setter methods, but anything scoped should be new and thus have the new codegen anyway)

	scopeFilter := req.GetScopeFilter()
	if err := scopes.ValidateFilter(scopeFilter); err != nil {
		return nil, "", trace.Wrap(err)
	}
	return s.service.ListResourcesWithFilter(ctx, int(req.GetPageSize()), req.GetPageToken(), func(foo *foov1.Foo) bool {
		return scopes.MatchScope(scopeFilter, foo.GetScope())
	})
}
```

> #### Range

Reviewer 4: Is a Range method always necessary? When do I need one?

Author: I think Range is more natural than List when consumed by the API layer, specifically when you need to filter listed resource with an authz check, which will always be necessary if the resources are scoped. With Range you allow the API layer to do the filtering instead of trying to pass a filter function down to the storage layer or re-page a paginated list

Reviewer 4: So should I favor Range over List, if I can, going forward?

I'm trying determine and, hopefully, cut down the amount of storage-level boilerplate that's needed for a new resource.

The Range method should accept the protobuf `List<Resource>Request` message as a parameter, so that options including filters can be added to the request in a backward-compatible manner.

A range operation should not abort entirely if a single item cannot be
converted from a `backend.Item`, it should instead be logged, and the rest of
the range should be processed.
Aborting an entire range when a single entry is invalid causes Teleport to be
permanently unhealthy since it is never able to load or cache the affected
resource(s).
`generic.(ScopeAware)ServiceWrapper` handles this.

```go
// RangeFoos ranges over all foos matching any scope filter specified in the
// request, between startKey and endKey interpreted as scoped resource cursors.

> func (s *FooService) RangeFoos(ctx context.Context, req *foov1.ListFoosRequest, startKey, endKey string) iter.Seq2[*foov1.Foo, error] {

Reviewer 4: This has the same limitations around composite keys that other string-based methods have.

Author: can you be specific, is this an ask to change the Range API? I don't see the limitation if you define a cursor format for your composite-key resource

Reviewer 4: Just an observation from someone who had to make a composite key work.

	scopeFilter := req.GetScopeFilter()
	if err := scopes.ValidateFilter(scopeFilter); err != nil {
		return stream.Fail[*foov1.Foo](trace.Wrap(err))
	}

> 	return stream.FilterMap(s.service.Resources(ctx, startKey, endKey), func(foo *foov1.Foo) (*foov1.Foo, bool) {

Reviewer 4: Should we push the filtering to the generic service instead?

Author: we _could_ push the scope filter into the generic service instead, and that will be a big improvement if it can convert some scope filters into key prefix queries. But filtering in the resource storage service is useful in case you want more filter options

		return foo, scopes.MatchScope(scopeFilter, foo.GetScope())
	})
}
```

### RBAC

Making use of the `stream.FilterMap` pattern can be used to verify access to a resource when listing:

```go
// ListFoos returns a page of Foo resources.
func (s *Service) ListFoos(ctx context.Context, req *foov1.ListFoosRequest) (*foov1.ListFoosResponse, error) {
	authzContext, err := s.cfg.ScopedAuthorizer.AuthorizeScoped(ctx)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	// do a pre-check to weed out requests that definitely won't be authorized.
	ruleCtx := authzContext.RuleContext()
	if err := authzContext.CheckerContext.CheckMaybeHasAccessToRules(&ruleCtx, foos.Kind, types.VerbReadNoSecrets, types.VerbList); err != nil {
		return nil, trace.Wrap(err)
	}

	// list method scope filters must use identity-based defaults per RFD 0229i

> 	req.SetScopeFilter(authzContext.CheckerContext.ResolveScopeFilter(req.GetScopeFilter()))

Reviewer 4: Should we pass this outside of the request, with a distinct type, so it's clearer it needs to be pre-calculated?

It's unusual to modify request fields this way.

Author: the goal of using the request as the parameter to List is that can we easily extend it with optional fields like filters without adding another method parameter every time. Is it specifically because the request is a protobuf type that you don't like modifying it? The whole question of this RFD amendment is if the team will accept passing the protobuf request to Get/Delete/List or demand something else (I would appreciate a concrete suggestion for the parameter types of Get/Delete/List if you would demand something else)

This line specifically I'm taking straight from a real implementation https://github.com/gravitational/teleport/pull/68320/changes#diff-41ff0ae1d040380457025c1174b7844587378270e9f450872c87ed33fbb2ae05R400-R401

Reviewer 4: I think the main downsides I see here is that this assumes gRPC and storage have identical interfaces.

In the case of scoped_filter, the implementer can forget to re-assign it via `authzContext.CheckerContext.ResolveScopeFilter`, meaning now it's forwarding some request parameter with unknown contents.

If storage needs some option that doesn't make sense to gRPC, or needs to be pre-calculated, gRPC will nevertheless expose the option to its API callers. Those fields now leak internals and can be incorrectly forwarded to storage if not explicitly handled. That will lead us to situations where a certain field exists but should never be set by gRPC callers.

As an alternative we could craft our Get/List/Delete "request/option" types in the storage layer and have them be part of the signature. Similar idea, but it doesn't assume gRPC and storage have identical interfaces.

> if you would demand something else

Demand is too strong of a word. This is more of a comment/suggestion.

Author: > I think the main downsides I see here is that this assumes gRPC and storage have identical interfaces.

Get and List _must_ have matching parameters between gRPC and storage if you want the resource to be cacheable, because the cache upstream might be a storage service or a gRPC client. Not necessarily identical, but the parameters must be 1-1 convertible between them if you are willing to wrap one upstream and convert the parameters to match the signature of the other.

> Should we pass this outside of the request, with a distinct type, so it's clearer it needs to be pre-calculated?

> In the case of scoped_filter, the implementer can forget to re-assign it via authzContext.CheckerContext.ResolveScopeFilter, meaning now it's forwarding some request parameter with unknown contents.

I do kinda like the idea of a separate type for a resolved (and validated?) scope filter, it's a question whether the benefits are worth the cost of an intermediate request type, maybe they are

> If storage needs some option that doesn't make sense to gRPC, or needs to be pre-calculated, gRPC will nevertheless expose the option to its API callers. Those fields now leak internals and can be incorrectly forwarded to storage if not explicitly handled. That will lead us to situations where a certain field exists but should never be set by gRPC callers.

In this case you are going beyond the baseline recommendations of RFD153, which is fine, you can have additional methods or parameters not explicitly recommended here. Don't put fields that shouldn't be set by clients into the proto request, I agree

> As an alternative we could craft our Get/List/Delete "request/option" types in the storage layer and have them be part of the signature. Similar idea, but it doesn't assume gRPC and storage have identical interfaces.

without rearchitecting all of teleport I think these would need to be defined in lib/services (or another package importable from there and everywhere else), you could have the storage service implement that interface, you'd wrap the gRPC client to match before passing it to the cache. I suppose it's doable you just end up converting back and forth between the proto request and the non-proto request

Reviewer 4: > Get and List must have matching parameters between gRPC and storage if you want the resource to be cacheable, because the cache upstream might be a storage service or a gRPC client

> without rearchitecting all of teleport I think these would need to be defined in lib/services

True. I think I see why you want to push the requests all the way. The distinction between a "storage service" and a "gRPC service" is indeed muddled in Teleport.

I still dislike the coding pattern of scope_filter, but I won't object to the protobuf requests. LGTM.

	stream := stream.FilterMap(

> 		s.cfg.Reader.RangeFoos(ctx, req, req.GetPageToken(), ""),

Reviewer 4: If Service.List doesn't use the Storage.List, then what does?

Author: the current cache implementation `genericLister.Range` needs an upstream List, and I think this makes sense because the upstream is likely to be a gRPC client that can easily implement list but not-so-easily implement Range

Reviewer 4: Gotcha. It probably sounds silly, but I would find it useful if we could mention that.

		func(foo *foov1.Foo) (*foov1.Foo, bool) {
			// Skip foos the caller is not authorized to see

> 			ruleCtx := authzContext.RuleContext()
> 			ruleCtx.Resource153 = foo

Reviewer 4: Maybe add this:

```suggestion
			ruleCtx := authzContext.RuleContextForResource(foo)
```

Author: not a bad suggestion but my goal here is to agree on the parameters for Get/Delete/List not to change authz implementation details

			if err := authzContext.CheckerContext.Decision(ctx, foo.GetScope(), func(checker *services.ScopedAccessChecker) error {
				return checker.CheckAccessToRules(&ruleCtx, foos.Kind, types.VerbReadNoSecrets, types.VerbList)
			}); err != nil {
				return nil, false
			}
			return foo, true
		},
	)
	page, nextPageToken, err := generic.CollectPageAndCursor(
		stream,
		int(req.GetPageSize()),
		foos.MakeCursor,

> 	)

Reviewer 4: This entire block is a lot of boilerplate. It has to be carefully tested too.

Could we provide a wrapper to simplify this?

Author: it's not a bad idea but also not my immediate goal, do to time constraints I won't be designing and implementing this at the moment

Reviewer 4: @rosstimothy, any chance we could get something like this in the pipeline? Writing a resource153 is already pretty painful/obtuse even without the scopes boilerplate in there.

func (p *fooParser) parse(event backend.Event) (types.Resource, error) {
	switch event.Type {
	case types.OpDelete:
		sqn, err := fooNameFromKey(event.Item.Key)
		if err != nil {
			return nil, trace.Wrap(err)
		}
		foo := foov1.Foo_builder{
			Kind:    foos.Kind,
			Version: types.V1,
			Metadata: headerv1.Metadata_builder{
				Name: sqn.Name,
			}.Build(),
			Scope: sqn.Scope,
		}.Build()
		return types.Resource153ToLegacy(foo), nil
	case types.OpPut:
		foo, err := services.UnmarshalProtoResource[*foov1.Foo](
			event.Item.Value,
			services.WithExpires(event.Item.Expires),
			services.WithRevision(event.Item.Revision),
		)
		if err != nil {
			return nil, trace.Wrap(err)
		}
		return types.Resource153ToLegacy(foo), nil
	default:
		return nil, trace.BadParameter("event %v is not supported", event.Type)
	}
}

> func fooNameFromKey(key backend.Key) (scopes.QualifiedName, error) {

Reviewer 4: Could we also provide a helper for this? Say, `SimpleNameFromScopedKey` for single-name resources.

	switch {
	case key.HasPrefix(fooScopedWatchPrefix()):
		components := key.TrimPrefix(fooScopedWatchPrefix()).Components()
		if len(components) != 2 {

> 			return scopes.QualifiedName{}, trace.NotFound("failed parsing %v", key.String())

Reviewer 3: IMO - since this is likely to be copy-pasted a fair bit, we should make sure these errors are a little more descriptive. E.G:

```suggestion
			return scopes.QualifiedName{}, trace.NotFound("expected 2 components, got %d parsing backend key %v", len(components), key.String())
```

all compatible versions of Teleport are using the new version of the API, the old API may be cleaned up.

### Reference Implementation

> A branch with a complete implementation of a Foo service including the proto specification, backend storage support, a gRPC API layer with authorization, `tctl` support, events support, and cache support can be found at https://github.com/gravitational/teleport/pull/68358

Reviewer 4: nit: full stop.

```suggestion
A branch with a complete implementation of a Foo service including the proto specification, backend storage support, a gRPC API layer with authorization, `tctl` support, events support, and cache support can be found at https://github.com/gravitational/teleport/pull/68358.
```
